### ISSUE TITLE
Dependent Dropdowns and Dynamic Fields

### ISSUE THREAD

**Issue by parlays** [jaredpalmer/formik#751]
Hello!  I have read through many of the issues and online examples but I have not found an example of using dynamic fields.

I would like to add or remove form fields based on the value of one of the dropdowns.  If the dropdown has "option A" selected, I would like to add two more dropdowns and populate them via an AJAX call.  If "option B" is selected, I would like to add an input box and remove the dropdowns displayed for "option A".

I tried using a custom react component but it is not clear how to add or remove form elements based on the value of another form element.  Thanks!

**Comment by SimpleCookie** [comment]
I am looking into the exact same thing right now. Still haven't found any answer.
I can't and shouldn't do setField inside render, otherwise that would be an option for me. But it's a major anti-pattern.

**Comment by SimpleCookie** [comment]
Okay @parlays I have solved the issue. It's not a pretty solution but it's a workaround.
I have a custom component which obviously calls `setFieldValue`, but it also calls a `onValueChangeCallback` to my container/page, which in turn updates the "options" prop to my form based on the option chosen.

To update the dynamic field's value, I am also calling `setFieldValue` on the `onChange` for my static field.

I'm sorry, I know my explanation is poor and hard to follow. My code and form are quite complex so it's difficult to give you a simple explanation.

But basically, what you want to do is to handle it through a callback I guess.

**Comment by pavanmehta91** [comment]
@SimpleCookie Can you give an example to show this?

**Comment by Chris-Boe** [comment]
@jaredpalmer Do you have any recommendations for dynamic and dependent fields?

**Comment by ajitfawade** [comment]
> @pavanmehta91 I'm sorry I don't have the possibility of creating an example for you right now.
> But .. Say you have 2 dropdowns, parent A and dependant child B.
> 
> When you change A, you want to update the options of B accordingly.
> What you need to do is, you need to first use `setfieldValue` to set value of A, and then you need to update the options of B, I did this using a callback. And then you also need to use `setfieldValue` to set the initial value of B.

Could you please give an example for the above case?

**Comment by dzg** [comment]
Any word on this? What's the most elegant way to solve?

**Comment by goldenshun** [comment]
I came across this looking for a way to set default fields in Formik's `values` when a conditional set of form fields is rendered. This is a simplified version of what my code looks like:

```
const ConditionalFields = (props) => {
  const { handleChange, setFieldValue, values } = props;

  useEffect(() => {
    // If my values are undefined, set them in Formik
    if (!values.myValue) {
      setFieldValue('myValue', 'default');
    }
  });

  // myValue is not defined so stop rendering, useEffect above will cause a re-render if necessary
  if (!values.myValue) return null;

  return (
    <div>
      <input type="text" value={values.myValue} onChange={handleChange} />
    </div>
  );
};
```

The premise of this solution is that you halt rendering when you encounter uninitialized fields, set field values in Formik's store, which then triggers a re-render. Not elegant, but could get you out of a bind.

**Comment by malyzeli** [comment]
@goldenshun Thanks for tip, I updated your solution a bit and here is my approach... 
- using `connect` helper to hook into form context instead of passing props explicitly
- using `getIn` helper to support nested field name syntax
- delegates rendering onto child `Field` component so it should be interchangeable

Component code:
```js
import { connect, Field, getIn } from "formik";
import { any, object, string } from "prop-types";
import React, { useEffect } from "react";

const OptionalField = ({
  formik: { values, setFieldValue },
  name,
  defaultValue,
  ...props
}) => {
  useEffect(() => {
    // set value if not defined
    if (getIn(values, name) === undefined) {
      setFieldValue(name, defaultValue);
    }
  });

  const value = getIn(values, name);

  // stop rendering, effect above will rerender
  if (value === undefined) return null;

  return <Field name={name} {...props} />;
};

OptionalField.propTypes = {
  formik: object.isRequired,
  name: string.isRequired,
  defaultValue: any,
};

OptionalField.defaultProps = {
  defaultValue: null,
};

export default connect(OptionalField);
```

Example usage:
```js
<OptionalField name="foo" defaultValue="bar" />

<OptionalField name="a.b" defaultValue={1} />

<OptionalField name="list" defaultValue={[1, 2, 3]} />

<OptionalField name="set[0]" defaultValue={true} />

<OptionalField name="set[1].value" defaultValue={{ a: 1, b: 2 }} />
```

Enjoy!

**Comment by goldenshun** [comment]
Super cool @malyzeli! I was unaware of the `connect` and `getIn` helpers ❤️

**Comment by malyzeli** [comment]
@goldenshun Yep, they are not documented well.. I found it by accident while browsing through the source code! :-D 

@jaredpalmer Can you please add page to API Reference listing available helpers? 
Maybe it's sufficient mentioning just `getIn`, since it's useful in situations like this. I see other helpers are being used internally, though I'm not sure if they could be reused for other purposes. But I believe you know better, so you could provide some hints!

**Comment by voliva** [comment]
I've made a solution inspired on @SimpleCookie: I created a component called FormSpy:

```tsx
import { Field, FieldProps } from 'formik';
import * as React from 'react';
import { useRef } from 'react';

interface FieldSpyProps {
    name: string;
    onChange: (value: any) => void;
}

const empty = Symbol('empty');
export const FieldSpy = ({ name, onChange }: FieldSpyProps) => {
    const value = useRef<any>(empty);

    return (
        <Field name={name}>
            {({ field }: FieldProps) => {
                if (value.current !== empty && value.current !== field.value) {
                    onChange(field.value);
                }
                value.current = field.value;

                return null;
            }}
        </Field>
    );
};
```

So now, whenever I want to "listen" to changes on a particular field in my form container, I can just:

```ts
            <Formik initialValues={defaultValue} onSubmit={onSubmit}>
                <Form>
                    <TextField name="username" />
                    <FieldSpy name="username" onChange={onUsernameChange} />
                </Form>
            </Formik>
```

What's your opinion on this? I haven't fully tried this, but it seems like it should cover most of the use cases.
The only problem though is that I can't make the call trigger on blur....

**Comment by ltfschoen** [comment]
I'd like to share how I solved a similar situation.

I had two selection box fields A and B using Formik.
Both A and B call a method that is passed to prop `getOptions` and list data that is retrieved
I conditionally rendered B only when the user selected an option in A (via `onNameSelect`) that returned
a value and stored it in `this.state.object.nameID`.
Each time I conditionally rendered B, in `onNameSelect` I reset selection box B so no value was chosen
(since each different Name may have a different list of Profiles available).
And each time a different option in A is chosen (which causes B to be conditionally be rendered), 
it runs `setFieldValue={() => setFieldValue("object.profileID", "0", false)}`, which
appears to re-render selection box B, and triggers a call to `getProfileOptions` and returns a
list of profiles in the selection box B that correspond to the current `this.state.object.nameID`.

``
...
getNameOptions = (search, callbackFunc) => {
  NameStore.list((999), resp => {
    const options = resp.result.map((n, i) => { return { label: n.name, value: n.id } });
    callbackFunc(options);
  });
}

getProfileOptions = (search, callbackFunc) => {
  // Only fetch Profiles associated with the Name that the user must have chosen first.
  if (this.state.object === undefined || this.state.object.nameID === undefined) {
    callbackFunc([]);
    return;
  }

  ProfileStore.list(this.state.object.nameID, resp => {
    const options = resp.result.map((p, i) => { return { label: p.name, value: p.id } });
    this.setState({ loading: false });
    callbackFunc(options);
  });
}

onNameSelect = (inputValue) => {
  if (!this.state.object.nameID !== inputValue.id)) {
    let object = this.state.object;
    // Reset profileID when nameID changes
    object.profileID = null;
    object.nameID = v.value;
    this.setState({
      object,
    });
  }
}

onProfileSelect = (inputValue) => {
  if (this.state.object.profileID !== inputValue.id)) {
    let object = this.state.object;
    object.profileID = v.value;
    this.setState({
      object
    });
  }
}

...

<Field
  id="nameID"
  name="object.nameID"
  type="text"
  value={values.object.nameID}
  onChange={this.onNameChange}
  getOptions={this.getNameOptions}
  // Hack: trigger Profile ID list to refresh
  // whenever the Name ID changes
  setFieldValue={() => setFieldValue("object.profileID", "0", false)}
  ...
/>

{values.object.nameID &&
  <Field
    id="profileID"
    name="object.profileID"
    type="text"
    value={values.object.profileID}
    onChange={this.onProfileSelect}
    getOptions={this.getProfileOptions}
    ...
  />
}
```

**Comment by Maskedman99** [comment]
Found a solution here 
https://stackoverflow.com/questions/60517777/how-can-i-create-connected-dependent-select-elements-in-formik

The code sandbox in the solution 
https://codesandbox.io/embed/formik-example-3jfxh?fontsize=14&hidenavigation=1&theme=dark

**Comment by richAtreides** [comment]
> 

Is name the field to which it is connected? trying to piece this together. Thanks!
